### PR TITLE
Tmpltbank params update rebase

### PR DESCRIPTION
**Pull request by spxiwh**
This patch makes a number of changes to the internals of the template bank module. The basic goal here is to update the PN terms used when doing geometric placement (specifically spin terms at >= 2.5PN order), but also to keep pycbc up-to-date with lalsimulation so that changes to PN terms automatically propagate through. This comes at a cost though, having to call lasimulation many times and then map outputs to the large arrays is slow, and while I've done what I can to optimize, this does give a speed hit. For stochastic banks this can slow the code down by up to an order of magnitude, for the geometric bank the hit seems to be a factor of 2. Adding a parallel routine in C (either via weave or a wrapping function in lalsimulation) would be desirable in the future, but is *not* under the scope of this pull request. For now I have preserved the old routine for generating PN terms (which is much faster, but does not have updated spin terms).

I want to run another bank generation test on this before removing the work in progress label. But I post this now for feedback in the meantime.

Note that PN order is still an argument, and when *larger* PN terms become available (4PN, 4.5PN ...) one will still need to add lines into this code to deal with the new terms and place them correctly in the resulting matrices (ie. does 4PN have log terms, or loglog terms? ... 4PN is also a "special" order, as the non-log terms are degenerate with an overall timeshift and therefore must not be included here or it would make the matrix degenerate (as 2.5PN terms are also not used because they are a constant phase shift)).

### CODE REVIEW

**Comment by titodalcanton** [comment]
Hi Ian, that's one huge diff! I'll take a while to go through it, however here's one thing that struck my eye: could you find better names than `get_chirp_params_new`/`_old`? Maybe "`_hardcoded`" vs "`_lalsim`"? New and old are observer-dependent concepts :)

**Comment by spxiwh** [comment]
@titodalcanton I went deliberately with "old" and "new" because I want to remove "old" in the near future, but want the "new" function available now for subsequent O2 bank construction.

The only reason I don't completely remove "old" at this point, is that I want to keep it around as a benchmark when trying to speed up "new". I can change this to "harcoded" and "lalsim" if you want though.

**Comment by spxiwh** [comment]
Hi @titodalcanton. I've now checked that all the examples on the tmpltbank documentation page build, and they do with some changes to the codes *not* used in uberbank construction. The bank sizes in all cases do not change (other than small changes in the stochastic code output, which is expected as the random process is no longer identical) when PN order is less than 2.5. Bank sizes have small changes with larger PN orders, but this is on the few percent level. Finally, I've run this through the uberbank workflow and it's running fine with only small changes in resulting bank size.

I've also added some doc fixes from #1360 here to avoid any issues between the two changes.

Now I'm ready to remove the "work in progress" label.

File: bin/pycbc_geom_aligned_2dstack [hunk 1]
```diff
@@ -162,15 +162,13 @@ if len(evals) < 5:
...
 # mapping from xi_i to masses and spins
-rTotmass, rEta, rBeta, rSigma, rGamma, rSpin1z, rSpin2z = \
+rMass1, rMass2, rSpin1z, rSpin2z = \
       pycbc.tmpltbank.get_random_mass(2000000, massRangeParams)
-diff = (rTotmass*rTotmass * (1-4*rEta))**0.5
-rMass1 = (rTotmass + diff)/2.
-rMass2 = (rTotmass - diff)/2.
-rChis = (rSpin1z + rSpin2z)/2.
+rTotmass = rMass1 + rMass2
+rEta = (rMass1 * rMass2) / (rTotmass * rTotmass)
```
Comment (titodalcanton) [inline bin/pycbc_geom_aligned_2dstack]: Ian, I'd use pnutils.mass1_mass2_to_mtotal_eta() here, as you've done later in `pycbc_geom_nonspinbank`.

File: pycbc/tmpltbank/lambda_mapping.py [hunk 2]
```diff
@@ -187,16 +276,17 @@ def get_chirp_params(totmass, eta, beta, sigma, gamma, chis, f0, order):
     --------
     lambdas : list of floats or numpy.arrays
         The lambda coordinates for the input system(s)
-    """ %(pycbcValidOrdersHelpDescriptions)
+    """
+
+    totmass = mass1 + mass2
+    eta = mass1 * mass2 / (totmass * totmass)
```
Comment (titodalcanton) [inline pycbc/tmpltbank/lambda_mapping.py]: Same comment as before here, can you use the pnutils function?

File: test/test_tmpltbank.py [hunk 3]
```diff
@@ -403,21 +395,17 @@ def test_get_mass_distribution(self):
...
+        masses2 = [2.21,1.79,0.41,0.29]
+        xis1 = pycbc.tmpltbank.get_cov_params(masses1[0], masses1[1],
+                 masses1[2], masses1[3], self.metricParams, self.f_upper)
+        xis2 = pycbc.tmpltbank.get_cov_params(masses2[0], masses2[1],
+                 masses2[2], masses2[3], self.metricParams, self.f_upper)
 
         testXis = [xis1[0],xis1[1]]
-        bestMasses = masses2
+        b_mtot = masses2[0] + masses2[1]
+        b_eta = masses2[0] * masses2[1] / (b_mtot * b_mtot)
```
Comment (titodalcanton) [inline test/test_tmpltbank.py]: Same pnutils comment here.

File: test/test_tmpltbank.py [hunk 4]
```diff
@@ -470,20 +452,16 @@ def test_stack_xi_direction(self):
...
-                 self.metricParams, self.f_upper)
+                 masses1[2], masses1[3], self.metricParams, self.f_upper)
         xis2 = pycbc.tmpltbank.get_cov_params(masses2[0], masses2[1], \
-                 spinSet2[0], spinSet2[1], spinSet2[2], spinSet2[3], \
-                 self.metricParams, self.f_upper)
+                 masses2[2], masses2[3], self.metricParams, self.f_upper)
         testXis = [xis1[0],xis1[1]]
-        bestMasses = masses2
+        b_mtot = masses2[0] + masses2[1]
+        b_eta = masses2[0] * masses2[1] / (b_mtot * b_mtot)
```
Comment (titodalcanton) [inline test/test_tmpltbank.py]: Same pnutils comment here.

File: pycbc/tmpltbank/lambda_mapping.py [hunk 5]
```diff
@@ -153,25 +148,119 @@ def ethinca_order_from_string(order):
...
+    Returns
+    --------
+    lambdas : list of floats or numpy.arrays
+        The lambda coordinates for the input system(s)
+    """
+
+    # Determine whether array or single value input
+    sngl_inp = False
+    try:
+        num_trigs = len(mass1)
```
Comment (titodalcanton) [inline pycbc/tmpltbank/lambda_mapping.py]: Why "trigs" here? Would `num_templates` or `num_points` be more correct?

File: pycbc/tmpltbank/lambda_mapping.py [hunk 6]
```diff
@@ -153,25 +148,119 @@ def ethinca_order_from_string(order):
...
+        mass1 = numpy.array([mass1])
+        mass2 = numpy.array([mass2])
+        spin1z = numpy.array([spin1z])
+        spin2z = numpy.array([spin2z])
+        num_trigs = 1
+    lal_pars = CreateDict()
+    phasing_vs = numpy.zeros([num_trigs, 13])
+    phasing_vlogvs = numpy.zeros([num_trigs, 13])
+    phasing_vlogvsqs = numpy.zeros([num_trigs, 13])
+    for i in xrange(len(mass1)):
```
Comment (titodalcanton) [inline pycbc/tmpltbank/lambda_mapping.py]: You can use `num_trigs` here instead of the `len()`.

File: pycbc/tmpltbank/lambda_mapping.py [hunk 7]
```diff
@@ -153,25 +148,119 @@ def ethinca_order_from_string(order):
...
+        phasing = lalsimulation.SimInspiralTaylorF2AlignedPhasing(
+                            mass1[i], mass2[i], spin1z[i], spin2z[i], lal_pars)
+        phasing_vs[i] = phasing.v
+        phasing_vlogvs[i] = phasing.vlogv
+        phasing_vlogvsqs[i] = phasing.vlogvsq
+
+    phasing_vs = numpy.array(phasing_vs)
+    phasing_vlogvs = numpy.array(phasing_vlogvs)
+    phasing_vlogvsqs = numpy.array(phasing_vlogvsqs)
```
Comment (titodalcanton) [inline pycbc/tmpltbank/lambda_mapping.py]: Aren't `phasing_*` already Numpy arrays?

**Comment by spxiwh** [comment]
Patch pushed @titodalcanton

**Pull request merged**